pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,551 @@
+#pragma once
+
+#include <bit>
+#include <complex>
+#include <cstring>
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <mpi.h>
+#include <stdexcept>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+#include <vector>
+
+#include "Error.hpp"
+
+namespace Pennylane::MPI {
+inline void errhandler(int errcode, const char *str) {
+    char msg[MPI_MAX_ERROR_STRING];
+    int resultlen;
+    MPI_Error_string(errcode, msg, &resultlen);
+    fprintf(stderr, "%s: %s\n", str, msg);
+    MPI_Abort(MPI_COMM_WORLD, 1);
+}
+
+#define PL_MPI_IS_SUCCESS(fn)                                                  \
+    {                                                                          \
+        int errcode;                                                           \
+        errcode = (fn);                                                        \
+        if (errcode != MPI_SUCCESS)                                            \
+            errhandler(errcode, #fn);                                          \
+    }
+
+#define CPPTYPE(s) std::type_index(typeid(s))
+
+class MPIManager {
+  private:
+    bool isExternalComm_;
+    int rank_;
+    int size_per_node_;
+    int size_;
+    MPI_Comm communiator_;
+
+    std::string vendor_;
+    int version_;
+    int subversion_;
+
+  public:
+    MPIManager() : communiator_(MPI_COMM_WORLD) {
+        isExternalComm_ = true;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communiator_, &rank_));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communiator_, &size_));
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+    }
+
+    MPIManager(MPI_Comm communicator) : communiator_(communicator) {
+        isExternalComm_ = true;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communiator_, &rank_));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communiator_, &size_));
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+    }
+
+    MPIManager(int argc, char **argv) {
+        PL_MPI_IS_SUCCESS(MPI_Init(&argc, &argv));
+        isExternalComm_ = false;
+        communiator_ = MPI_COMM_WORLD;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communiator_, &rank_));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communiator_, &size_));
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+        check_mpi_config();
+    }
+
+    MPIManager(const MPIManager &other) {
+        isExternalComm_ = true;
+        rank_ = other.rank_;
+        size_ = other.size_;
+        communiator_ = other.communiator_;
+        vendor_ = other.vendor_;
+        version_ = other.version_;
+        subversion_ = other.subversion_;
+        size_per_node_ = other.size_per_node_;
+    }
+
+    ~MPIManager() {
```

Comment:
While I don't expect us to derive from this class, we should ensure the destructor is virtual. If explicitly not derivable the class itself can be declared `final` to avoid it being inherited from.

Reply:
done!